### ISSUE TITLE
RFE: semantics to allow closure calls to take place on non-closures

### ISSUE THREAD

**Issue by jwise** [google/jsonnet#477]
It would be really nice to be able to write something like `std.makeArray(8, 0.0)`, or -- more to the point -- `std.makeArray(4, std.makeArray(5, std.makeArray(5, 0.0)))`, rather than `std.makeArray(4, function(_) std.makeArray(5, function(_) std.makeArray(5, function(_) 0.0)))`.

The easiest way I see to make that work is to add an application rule for non-functions that `\rule{e_0 ↓ e_0', e_0' ∈ {Number, String, Array, Object, Boolean, null}}{e_0(e_1 ... e_m) ↓ e_0})` (or such).  This makes the language a little bit gross inasmuch as it makes static checking of it a little less useful, but on the other hand, it really does simplify the syntax for common uses of makeArray...

**Comment by sbarzowski** [comment]
> This makes the language a little bit gross inasmuch as it makes static checking of it a little less useful, but on the other hand, it really does simplify the syntax for common uses of makeArray...

Yes. I think a bigger problem is that it could be harder for humans to understand the code. I agree we should do something about readability of such `makeArray` calls and similar stuff. It seems to me that the problem could be solved with two simple additions to stdlib:

```
local const(x) = function(_) x,
         constArray(n, val) = makeArray(n, const(val));
```

And then you could do:
`std.constArray(4, std.constArray(5, std.constArray(5, 0.0)))`

@sparkprime What do you think?

**Comment by marcelocantos** [comment]
A more concise closure syntax would obviate this: `std.makeArray(4, \_ std.makeArray(5, \_ std.makeArray(5, \_ 0.0)))`.

**Comment by sparkprime** [comment]
Adding arrow functions (there's a request somewhere else for that) would also solve this.

**Comment by jwise** [comment]
I think a `std.constArray` function (or `std.times` I think is the name that people use for this sort of thing) would solve my problem effectively.

**Comment by marcelocantos** [comment]
I'd argue for not adding single-purpose solutions like `std.const` and `std.constArray` if a general solution for more concise functions is on the cards.

My preference for `\x 1/x` is because it can be parsed (by humans and computers) left-to-right, reads naturally as "lambda x is one-on-x" and is about as concise as you can make it while retaining good readability. That said, arrow functions wouldn't be terrible: `x => 1/x`. They do get a bit noisier with curried functions, though: `\x \y \z x*y+z` vs `x => y => z => x*y+z`.

**Comment by sbarzowski** [comment]
I think even with concise lambdas there are _small_ advantages to adding these.

* `std.constArray`/`std.times` - We can later add a builtin implementation, which will be much faster than `std.makeArray(10000, \_ 42)`. Also it avoids the function argument completely, so less indirection.
* `std.const` - perhaps more readable. See `const(42)` vs `\_ 42`.

Both seem like a fine addition to the library to me either way, but I don't feel very strongly about it.

**Comment by marcelocantos** [comment]
Agreed that the special helpers would be a bit cleaner, but:

* I question the performance argument. Surely a const-detector could do a trivial rewrite of `std.makeArray(10, \_ expr)` as `std.$constArray(10, expr)`?
* The readability argument only holds, imo, because `\_ …` is currently an unfamiliar form.

While I wouldn't consider it a travesty to add the extra functions, I would put the question: if concise lambdas were already a thing, would we still feel the need for `constArray` and `const`?

**Comment by sparkprime** [comment]
The analysis to recognize `unused => expr` is indeed easy and my preference is to use the existing syntax of Javascript rather than functional languages or something custom.  I might also be persuaded to allow [3] * 10 because Python has it and it seems pretty harmless / low complexity cost.

**Comment by marcelocantos** [comment]
+1 for `[3] * 10` and don't forget `[1, 2, 3] * 10`.

**Comment by sparkprime** [comment]
`[1, 2, 3]` would come for free with `[3]` :)

**Comment by jwise** [comment]
I think `[3] * 10` would be just fine by me, and would solve my original problem very well -- `std.times` is what I was looking for, but the sugar for it is even better.  The way I would end up writing that would be something like `[[[0.0] * 5] * 5] * 4]`, which is very clear in intent.
